File: benchmark/autolifter/single-pass/mas.f
Config ExtraGrammar = "AutoLifter";

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

/*User provided programs*/

max = \a: Int. \b: Int. if < a b then b else a;

mas = fix (
  \f: Int -> Int -> List -> Int. \p: Int. \np: Int. \xs: List.
  match xs with
    nil _ -> max p np
  | cons {h, t} ->
    max (max p np) (f (+ (max np 0) h) (- (max p 0) h) t)
  end
) 0 0;

expected = \xs: List.
let res = (fix (
  \f: Int -> Int -> List -> {Int, Int}. \p: Int. \np: Int. \xs: List.
  match xs with
    nil _ -> {max p 0, max np 0}
  | cons {h, t} -> (f (+ (max np 0) h) (- (max p 0) h) t)
  end
) 0 0 xs) in
  max res.1 res.2;

main = single_pass mas;

File: ml-benchmark/autolifter/single-pass/mas.ml
type l = Cons of int * l | Nil of unit

let single_pass v =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let max a b = if a < b then b else a

let mas =
  let rec f p np xs =
    match xs with
    | Nil _ -> max p np
    | Cons (h, t) ->
        max (max p np) (f ((max np 0) + h) ((max p 0) - h) t)
  in
  f 0 0

let expected xs =
  let res =
    let rec f p np xs =
      match xs with
      | Nil _ -> (max p 0, max np 0)
      | Cons (h, t) ->
          f ((max np 0) + h) ((max p 0) - h) t
    in
    f 0 0 xs
  in
  match res with
  | a, b -> max a b

let program = single_pass mas